### PR TITLE
Fixed #9963: Display explicit error in case UID has no actual username

### LINKED ISSUE #9963: Don't crash when a getpass.getuser() call fails
**Apache Airflow version**: 1.10.11

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`): 1.18.4

**Environment**:

- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release): Ubuntu 20.04
- **Kernel** (e.g. `uname -a`): 
- **Install tools**:
- **Others**:

**What happened**:

There are many call occurrences to `getpass.getuser()` in the code. None are fenced with a try/except to protect them from failures. Though, in a containerized environment, using a process with an uid that is not defined in `/etc/passwd` is not uncommon when using a existing image. This makes airflow crash with errors like the one below, though the call to `getpass.getuser()` serves no critical purposes (it is sometimes used to annotate metrics...) AFAICT.

```
Traceback (most recent call last):
  File "/usr/local/bin/airflow", line 37, in <module>
    args.func(args)
  File "/usr/local/lib/python3.8/site-packages/airflow/utils/cli.py", line 76, in wrapper
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/airflow/bin/cli.py", line 1189, in scheduler
    job = jobs.SchedulerJob(
  File "<string>", line 4, in __init__
  File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/state.py", line 433, in _initialize_instance
    manager.dispatch.init_failure(self, args, kwargs)
  File "/usr/local/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.raise_(
  File "/usr/local/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
    raise exception
  File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/state.py", line 430, in _initialize_instance
    return manager.original_init(*mixed[1:], **kwargs)
  File "/usr/local/lib/python3.8/site-packages/airflow/jobs/scheduler_job.py", line 387, in __init__
    super(SchedulerJob, self).__init__(*args, **kwargs)
  File "<string>", line 6, in __init__
  File "/usr/local/lib/python3.8/site-packages/airflow/jobs/base_job.py", line 92, in __init__
    self.unixname = getpass.getuser()
  File "/usr/local/lib/python3.8/getpass.py", line 169, in getuser
    return pwd.getpwuid(os.getuid())[0]
KeyError: 'getpwuid(): uid not found: 1000'
```
I think i many cases it could be simply replaced with:
```python
try:
    <var> = getpass.getuser()
except KeyError:
    <var> = str(os.getuid())
```

**What you expected to happen**:

I expected Airflow to not crash because the uid is not mapped to a user defined in `/etc/passwd`

**How to reproduce it**:

Deploy the official airflow image in docker with `--user=9999` (or any other uid that is not 50000) or in k8s with a security context like:
```
securityContext:
  runAsUser: 9999
```
**Anything else we need to know**:

It even looks like the value returned by `getuser()` is hardly used, e.g. jobs see their unixname attribute assigned the value, but use of that attribute seem hardly critical:
```
 grep -HInr "\.unixname" "${VIRTUAL_ENV}"/lib/python3.8/site-packages/airflow{,_exporter}
airflow/migrations/versions/e3a246e0dc1_current_schema.py:105:            sa.Column('unixname', sa.String(length=1000), nullable=True),
airflow/migrations/versions/e3a246e0dc1_current_schema.py:166:            sa.Column('unixname', sa.String(length=1000), nullable=True),
airflow/migrations/versions/6e96a59344a4_make_taskinstance_pool_not_nullable.py:71:    unixname = Column(String(1000))
airflow/jobs/base_job.py:67:    unixname = Column(String(1000))
airflow/jobs/base_job.py:92:        self.unixname = getpass.getuser()
airflow/www/views.py:2809:        'unixname', 'hostname', 'start_date', 'end_date', 'latest_heartbeat')
airflow/www/views.py:3015:        'unixname', 'priority_weight', 'queue', 'queued_dttm', 'try_number',
airflow/www_rbac/views.py:2491:                    'executor_class', 'hostname', 'unixname']
airflow/www_rbac/views.py:2494:                      'hostname', 'unixname']
airflow/www_rbac/views.py:2689:                    'unixname', 'priority_weight', 'queue', 'queued_dttm', 'try_number',
airflow/models/taskinstance.py:157:    unixname = Column(String(1000))
airflow/models/taskinstance.py:201:        self.unixname = getpass.getuser()
airflow/models/taskinstance.py:487:            self.unixname = ti.unixname
```
(there is a `run_as_user` attribute that is)

### PR DESCRIPTION
**Pull request by andrewgodwin**
Previously, we used getpass.getuser() with no fallback, which errors out if there is no username specified for the current UID (which happens a lot more in environments like Docker & Kubernetes).

This updates most calls to use our own copy which has a fallback to return the UID as a string if there is no username. A couple of places are untouched as they are looking for an actual username as a connection default.

closes: #9963

### CODE REVIEW

**Review by potiuk — CHANGES_REQUESTED** [review]
I think this is quite wrong if the user is not defined. Not having the HOME/USER_ID defined is a problem of how the image is run, or rather how the image is prepared to cope with arbitrary user id. There are many more problems if you run airflow with arbitrary user id, such as access to write lgos or read dags - depending on how your deployment is done.

With the official Airflow Image, we are OpenShift-compatible, which allows to run the image with an arbittrary UID. We are further strengthening this with #15162 which I hope will be merged before 2.0.2 and in those cases the user entry is automatically created when arbitrary user id is used. 

I think trying to fix part of the problem with userpass is wrong. Rather than allowing a username-less UID,  we should be very explicit when this problem happens and explictly  error with appropriate error message ("please make sure the USER with UID=X is created, it has the "/etc/passwd" entry and appropriate HOME directory created and it has access to read DAGs and write to  LOGS dir." or similar.  Otherwise we are just passing a "hot-potato" problem elsewhere rather than face it and guide the user to fix it.

I think having a user "defined" when you use arbitrary UID is pretty much a standard POSIX expectation. There are many 3rd-party dependencies that expects the user to be well-defined (includin HOME_DIR) being available (for example to store cache there). Example gcloud/awscli/axzure - all of them store configuration/cache etc. in HOME_DIR. And by allowing non-POSIX-compliant behavior we are simply avoiding the problem and passing it to others rather than facing it.

**Comment by andrewgodwin** [comment]
Honestly, I agree, I was just following along with the original ticket's suggestion since nobody suggested an explicit error instead so I presumed there was implicit acceptance. I can update the PR to raise an exception instead, but leave the indirection via `utils.platform` in order to have a single place the error is caught/raised.

Generally, though, I _have_ found that a surprising number of tools will happily work  without an actual user record, so there is a user expectation here when they're firing things into Docker. I think it doesn't make sense to match that given Airflow is relatively dependent on user env/environment variables, but I'll see if I can make an error message that is nice enough to stop duplicate issues like #9963 being opened.

**Comment by potiuk** [comment]
> Generally, though, I have found that a surprising number of tools will happily work without an actual user record, so there is a user expectation here when they're firing things into Docker. I think it doesn't make sense to match that given Airflow is relatively dependent on user env/environment variables, but I'll see if I can make an error message that is nice enough to stop duplicate issues like #9963 being opened.

I do not think this is the user expectation. Simply the users are unaware this is a problem and it's better to educate them.
Much more surprising to the users, is that = they find that SOME of their dependencies do not work, and then it usually will end up with `airflow` being the culprit and issues raised towards the airflow community. I prefer much more to fail hard in this case and display an explicit error message rather than "pretend nothing happens". The cost of fixing such problem is directly proportional to the time it is needed to detect and diagnose the problem. The sooner we detect it and the more explicit we get, the less it cost for the user and the community to fix it. It's very low cost to detect the problem immediately when you run 'airflow'. It is far more costly to fix it when you find out after few months of running the DAGs that some of your tools misbehaved but you have not noticed it.

**Review by potiuk — APPROVED** [review]
Much nicer!

**Comment by potiuk** [comment]
@andrewgodwin -> One more change is needed. Imports are failing when you import providers. Providers are supposed to work even with older versions of Airflow, so you shoudl make it fall-back to the original imports from Airflow.

**Comment by andrewgodwin** [comment]
> Imports are failing when you import providers. Providers are supposed to work even with older versions of Airflow

Ah yes! That's on me, I'm still getting used to the needs of the different codebase portions. I'll get that in tomorrow.

File: airflow/utils/platform.py [hunk 1]
```diff
@@ -57,3 +58,21 @@ def get_airflow_git_version():
         log.debug(e)
 
     return git_version
+
+
+def getuser() -> str:
+    """
+    Gets the username associated with the current user, or a fallback
+    representation if there is no username (stringified UID)
```
Comment (uranusjr) [inline airflow/utils/platform.py]: Need to be update this to match the actual implementation.

Edit: Some *Note - we don't fallback to os.getuid() here* comments above seem out of place as well.

**Comment by andrewgodwin** [comment]
Any ideas why the static checks are failing on a problem that doesn't exist?

```tests/task/task_runner/test_standard_task_runner.py:145:52: F821 undefined name 'getpass'```

**Comment by uranusjr** [comment]
I think this is the issue: https://github.com/apache/airflow/pull/15212/files#diff-696345b1aad17ad1ac269c2224081205542b6a662c6d4a9ae731fb635eaa558eR145

**Comment by andrewgodwin** [comment]
> I think you missed one `getpass.getuser()` here:

Ah, hoisted by my own petard! It's my other diff sneaking in!

**Review by kaxil — APPROVED** [review]

**Pull request merged**